tests/resources/test_file_link.py

```diff
@@ -95,7 +95,59 @@ def test_upload(mock_isfile, mock_stat, mock_open, mock_boto3_client, collection
     mock_open.return_value.__enter__.return_value = 'Random file contents'
     mock_boto3_client.return_value = FakeS3Client({'VersionId': '3'})
 
-    dest_name = 'foo.txt'
+    dest_names = ['foo.txt', 'foo.TXT']  # Verify that capitalization in extension is fine
+    file_id = '12345'
+    version = '13'
+
+    # This is the dictionary structure we expect from the upload completion request
+    file_info_response = {
+        'file_info': {
+            'file_id': file_id,
+            'version': version
+        }
+    }
+    uploads_response = {
+        's3_region': 'us-east-1',
+        's3_bucket': 'temp-bucket',
+        'temporary_credentials': {
+            'access_key_id': '1234',
+            'secret_access_key': 'abbb8777',
+            'session_token': 'hefheuhuhhu83772333',
+        },
+        'uploads': [
+            {
+                's3_key': '66377378',
+                'upload_id': '111',
+            }
+        ]
+    }
+
+    for dest_name in dest_names:
+        session.set_responses(uploads_response, file_info_response)
+        file_link = collection.upload(dest_name)
+
+        url = 'projects/{}/datasets/{}/files/{}/versions/{}'\
+            .format(collection.project_id, collection.dataset_id, file_id, version)
+        assert file_link.dump() == FileLink(dest_name, url=url).dump()
```

Comment:
How is this testing that the extension is correctly read? I don't see any check of the `upload_json` structure. Or is the test that this simple occurs without throwing an exception, because previously there would have been an exception on `mimetypes.types_map['TXT']`?

Reply:
Yes, it's just a test that capitalization no longer fails.  I could add an `assert mimetype1 == mimetype2` test following if explicit case insensitivity makes more sense, but partly this is a band-aid to better typing approaches.

Reply:
I dove in, and (presuming I'm not missing anything) implementing that test would involve a bunch of dark art at the least.